arrow/src/main/java/org/apache/iceberg/arrow/vectorized/parquet/VectorizedDeltaLengthByteArrayValuesReader.java

```diff
@@ -0,0 +1,178 @@
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.arrow.vectorized.parquet;
+
+import java.io.EOFException;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.UUID;
+import java.util.function.IntUnaryOperator;
+import org.apache.arrow.vector.FieldVector;
+import org.apache.arrow.vector.IntVector;
+import org.apache.iceberg.arrow.ArrowAllocation;
+import org.apache.iceberg.io.CloseableGroup;
+import org.apache.parquet.bytes.ByteBufferInputStream;
+import org.apache.parquet.io.ParquetDecodingException;
+import org.apache.parquet.io.api.Binary;
+
+public class VectorizedDeltaLengthByteArrayValuesReader
+    implements VectorizedValuesReader, AutoCloseable {
+
+  private final VectorizedDeltaEncodedValuesReader lengthReader;
+  private final CloseableGroup closeables;
+
+  private ByteBufferInputStream in;
+  private IntVector lengthsVector;
+  private ByteBuffer byteBuffer;
+
+  VectorizedDeltaLengthByteArrayValuesReader() {
+    lengthReader = new VectorizedDeltaEncodedValuesReader();
+    closeables = new CloseableGroup();
+  }
+
+  @Override
+  public void initFromPage(int valueCount, ByteBufferInputStream inputStream) throws IOException {
+    lengthsVector = new IntVector("length-" + UUID.randomUUID(), ArrowAllocation.rootAllocator());
+    closeables.addCloseable(lengthsVector);
+    lengthReader.initFromPage(valueCount, inputStream);
+    lengthReader.readIntegers(lengthReader.getTotalValueCount(), lengthsVector, 0);
+    this.in = inputStream.remainingStream();
+  }
+
+  @Override
+  public Binary readBinary(int len) {
+    readValues(1, null, 0, x -> len, (f, i, v) -> byteBuffer = v);
+    return Binary.fromReusedByteBuffer(byteBuffer);
+  }
+
+  @Override
+  public void readBinary(int total, FieldVector vec, int rowId) {
+    readValues(
+        total,
+        vec,
+        rowId,
+        x -> lengthsVector.get(x),
+        (f, i, v) -> f.getDataBuffer().setBytes(i, v));
+  }
+
+  private void readValues(
+      int total,
+      FieldVector vec,
+      int rowId,
+      IntUnaryOperator getLength,
+      BinaryOutputWriter outputWriter) {
+    ByteBuffer buffer;
+    int length;
+    for (int i = 0; i < total; i++) {
+      length = getLength.applyAsInt(rowId + i);
+      try {
+        buffer = in.slice(length);
+      } catch (EOFException e) {
+        throw new ParquetDecodingException("Failed to read " + length + " bytes");
+      }
+      outputWriter.write(vec, rowId + i, buffer);
```

Comment:
I think there is a problem here: `rowId + i` is a row index, but the writer treats it as a byte offset. I suspect there might be data corruption.